File: frontend/src/services/notifications.ts
import api from './api'
import type { Notification, NotificationListResponse } from '../types/notification'

export async function fetchNotifications(page = 1, pageSize = 20): Promise<NotificationListResponse> {
  const { data } = await api.get<NotificationListResponse>('/api/v1/notifications/', {
    params: { page, page_size: pageSize },
  })
  return data
}

export async function fetchUnreadNotificationCount(): Promise<number> {
  const { data } = await api.get<{ count: number }>('/api/v1/notifications/unread-count')
  return data.count
}

export async function markNotificationRead(notificationId: number): Promise<Notification> {
  const { data } = await api.patch<Notification>(`/api/v1/notifications/${notificationId}/read`)
  return data
}

export async function markAllNotificationsRead(): Promise<void> {
  await api.patch('/api/v1/notifications/read-all')
}

File: backend/app/api/v1/notifications.py
from math import ceil

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.core.dependencies import get_current_user
from app.db.session import get_db
from app.models.issue import Issue
from app.models.notification import Notification
from app.models.user import User
from app.schemas.notification import (
    NotificationListResponse,
    NotificationResponse,
    UnreadCountResponse,
)
from app.services.notification_service import (
    get_unread_count,
    list_notifications,
    mark_all_as_read,
    mark_as_read,
)

router = APIRouter(prefix="/api/v1/notifications", tags=["notifications"])


@router.get("/", response_model=NotificationListResponse)
def get_notifications(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    items, total = list_notifications(db, current_user.id, page, page_size)
    return {
        "items": items,
        "total": total,
        "page": page,
        "page_size": page_size,
        "total_pages": ceil(total / page_size) if total else 0,
    }


@router.get("/unread-count", response_model=UnreadCountResponse)
def get_notification_unread_count(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return {"count": get_unread_count(db, current_user.id)}


@router.patch("/{notification_id}/read", response_model=NotificationResponse)
def mark_notification_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    notification = (
        db.query(Notification)
        .filter(Notification.id == notification_id, Notification.user_id == current_user.id)
        .first()
    )
    if not notification:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found.")
    notification = mark_as_read(db, notification)
    issue_title = (
        db.query(Issue.title).filter(Issue.id == notification.issue_id).scalar()
        if notification.issue_id
        else None
    )
    return {**notification.__dict__, "issue_title": issue_title}


@router.patch("/read-all")
def mark_notifications_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    updated = mark_all_as_read(db, current_user.id)
    return {"updated": updated}
